File: DSA/DivyanshKumarSingh_590012175/Day_39_Question_3.c
#include <stdio.h>

#define MAX 100

int arr[MAX], top1 = -1, top2 = MAX;

void push1(int x)
{
    if (top1 + 1 < top2)
    {
        arr[++top1] = x;
        printf("Pushed %d to Stack1\n", x);
    }
    else
    {
        printf("Stack Overflow in Stack1\n");
    }
}

void push2(int x)
{
    if (top1 + 1 < top2)
    {
        arr[--top2] = x;
        printf("Pushed %d to Stack2\n", x);
    }
    else
    {
        printf("Stack Overflow in Stack2\n");
    }
}

void pop1()
{
    if (top1 >= 0)
    {
        printf("Popped %d from Stack1\n", arr[top1--]);
    }
    else
    {
        printf("Stack Underflow in Stack1\n");
    }
}

void pop2()
{
    if (top2 < MAX)
    {
        printf("Popped %d from Stack2\n", arr[top2++]);
    }
    else
    {
        printf("Stack Underflow in Stack2\n");
    }
}

void print1()
{
    printf("Stack1: ");
    if (top1 == -1)
    {
        printf("Empty");
    }
    else
    {
        for (int i = 0; i <= top1; i++)
            printf("%d ", arr[i]);
    }
    printf("\n");
}

void print2()
{
    printf("Stack2: ");
    if (top2 == MAX)
    {
        printf("Empty");
    }
    else
    {
        for (int i = top2; i < MAX; i++)
            printf("%d ", arr[i]);
    }
    printf("\n");
}

int main()
{
    int n, op, st, x;

    printf("Enter number of operations: ");
    scanf("%d", &n);

    printf("\nInstructions:\n");
    printf("To PUSH: Enter 1 <stack_number> <value>\n");
    printf("To POP:  Enter 2 <stack_number>\n\n");

    for (int i = 0; i < n; i++)
    {
        printf("Operation %d: ", i + 1);
        scanf("%d", &op);
        if (op == 1)
        {
            scanf("%d %d", &st, &x);
            if (st == 1)
                push1(x);
            else if (st == 2)
                push2(x);
            else
                printf("Invalid stack number!\n");
        }
        else if (op == 2)
        {
            scanf("%d", &st);
            if (st == 1)
                pop1();
            else if (st == 2)
                pop2();
            else
                printf("Invalid stack number!\n");
        }
        else
        {
            printf("Invalid operation code!\n");
        }
    }

    printf("\nFinal Stack Status:\n");
    print1();
    print2();

    return 0;
}

File: DSA/YashKrishanGupta_590014798/Day_39_Question_3.java
public class Day_39_Question_3 {
    static class DualStack {
        int[] arr;
        int top1, top2;

        public DualStack(int n) {
            arr = new int[n];
            top1 = -1;
            top2 = n;
        }

        public void push1(int val) {
            if (top1 + 1 < top2) {
                arr[++top1] = val;
            } else {
                throw new RuntimeException("Stack Overflow");
            }
        }

        public void push2(int val) {
            if (top1 + 1 < top2) {
                arr[--top2] = val;
            } else {
                throw new RuntimeException("Stack Overflow");
            }
        }

        public int pop1() {
            if (top1 >= 0) {
                return arr[top1--];
            }
            throw new RuntimeException("Stack 1 Underflow");
        }

        public int pop2() {
            if (top2 < arr.length) {
                return arr[top2++];
            }
            throw new RuntimeException("Stack 2 Underflow");
        }

        public boolean isFull() {
            return top1 + 1 == top2;
        }

        public void printStack1() {
            if (top1 < 0) {
                System.out.println("Stack 1 is empty");
                return;
            }
            System.out.print("Stack 1 (bottom → top): ");
            for (int i = 0; i <= top1; i++) {
                System.out.print(arr[i] + " ");
            }
            System.out.println();
        }

        public void printStack2() {
            if (top2 >= arr.length) {
                System.out.println("Stack 2 is empty");
                return;
            }
            System.out.print("Stack 2 (bottom → top): ");
            for (int i = arr.length - 1; i >= top2; i--) {
                System.out.print(arr[i] + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        DualStack st = new DualStack(50);
        st.push1(1);
        st.push1(2);
        st.push2(8);
        st.push2(9);
        st.printStack1();
        st.printStack2();
        st.push1(3);
        st.pop2();
        st.printStack1();
        st.printStack2();
    }
}
